src/musig.rs

```diff
@@ -0,0 +1,1158 @@
+    }
+}
+
+/// An individual MuSig public nonce. Not to be confused with [`AggregatedNonce`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct PublicNonce(ffi::MusigPubNonce);
+
+impl CPtr for PublicNonce {
+    type Target = ffi::MusigPubNonce;
+
+    fn as_c_ptr(&self) -> *const Self::Target { self.as_ptr() }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target { self.as_mut_ptr() }
+}
+
+impl PublicNonce {
+    /// Serialize a PublicNonce
+    pub fn serialize(&self) -> [u8; ffi::MUSIG_PUBNONCE_SERIALIZED_LEN] {
+        let mut data = [0; ffi::MUSIG_PUBNONCE_SERIALIZED_LEN];
+        unsafe {
+            if ffi::secp256k1_musig_pubnonce_serialize(
+                ffi::secp256k1_context_no_precomp,
+                data.as_mut_ptr(),
+                self.as_ptr(),
+            ) == 0
+            {
+                // Only fails when the arguments are invalid which is not possible in safe rust
+                unreachable!("Arguments must be valid and well-typed")
+            } else {
+                data
+            }
+        }
+    }
+
+    /// Deserialize a PublicNonce from a portable byte representation
+    ///
+    /// # Errors:
+    ///
+    /// - MalformedArg: If the [`PublicNonce`] is 132 bytes, but out of curve order
+    pub fn from_byte_array(data: &[u8; ffi::MUSIG_PUBNONCE_SERIALIZED_LEN]) -> Result<Self, ParseError> {
+        let mut pub_nonce = MaybeUninit::<ffi::MusigPubNonce>::uninit();
+        unsafe {
+            if ffi::secp256k1_musig_pubnonce_parse(
+                ffi::secp256k1_context_no_precomp,
+                pub_nonce.as_mut_ptr(),
+                data.as_ptr(),
+            ) == 0
+            {
+                Err(ParseError::MalformedArg)
+            } else {
+                Ok(PublicNonce(pub_nonce.assume_init()))
+            }
+        }
+    }
+
+    /// Get a const pointer to the inner PublicNonce
+    pub fn as_ptr(&self) -> *const ffi::MusigPubNonce { &self.0 }
+
+    /// Get a mut pointer to the inner PublicNonce
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::MusigPubNonce { &mut self.0 }
+}
+
+/// Musig aggregated nonce computed by aggregating all individual public nonces
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct AggregatedNonce(ffi::MusigAggNonce);
+
+impl CPtr for AggregatedNonce {
+    type Target = ffi::MusigAggNonce;
+
+    fn as_c_ptr(&self) -> *const Self::Target { self.as_ptr() }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target { self.as_mut_ptr() }
+}
+
+impl AggregatedNonce {
+    /// Combine received public nonces into a single aggregated nonce
+    ///
+    /// This is useful to reduce the communication between signers, because instead
+    /// of everyone sending nonces to everyone else, there can be one party
+    /// receiving all nonces, combining the nonces with this function and then
+    /// sending only the combined nonce back to the signers.
```

Comment:
In dd26c3067044a89c503a32079c24227d6e1afca5:

In the upstream docs there's an extra line clarifying that if the aggregator lies, the signature will just be invalid. We should copy that here. (Will do in followup.)